pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -787,3 +789,108 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
                     check_mat[i][j] = check_mat[i][j-1]
 
     return OneDimensionalArray(seq1._dtype, check_mat[row][col][-1])
+
+def upper_bound(array, start, end, value, comp):
+    """
+    Finds the index of the first occurence of an element greater than value according to an order defined,in the given sorted OneDimensionalArray
+
+    Parameters
+    ========
+    array: OneDimensionalArray
+    The sorted array (sorted according to a custom comparator function) in which the upper bound has to be found
+
+    start: int
+    The staring index of the portion of the array in which the upper bound of a given value has to be looked for
+
+    end: int
+    The ending index of the portion of the array in which the upper bound of a given value has to be looked for
+
+    Returns
+    =======
+
+    output: int
+    Upper bound of the given value in the given sorted OneDimensionalArray
+
+    Examples
+    ========
+    >>> from pydatastructs import upper_bound, OneDimensionalArray as ODA
+    >>> arr = ODA(int, [4, 5, 5, 6, 7])
+    >>> upperBound = upper_bound(arr, 0, 4, 5, None)
+    >>> upperBound
+    3
+    >>> arr = ODA(int, [7, 6, 5, 5, 4])
+    >>> upperBound = upper_bound(arr, 0, 4, 5, lambda x, y: x >= y)
+    >>> upperBound
+    4
+
+    Note
+    ====
+
+    The OneDimensionalArray must be sorted beforehand
+    """
+    if comp is None:
+        comp = lambda a, b: (a <= b)
+    index = end
+    if not comp(array[start], value):
+        index = start
+    while start <= end:
+        mid = (start+end)//2
+        if comp(array[mid],value):
+            start = mid + 1
+        else:
+            index = mid
+            end = mid -1
+    return index
+
+def lower_bound(array, start, end, value, comp):
+    """
+    Finds the the index of the first occurence of an element which is not less than value according to an order defined, in the given OneDimensionalArray
+
+    Parameters
+    ========
+    array: OneDimensionalArray
+    The sorted array (sorted according to a custom comparator function) in which the lower bound has to be found
+
+    start: int
+    The staring index of the portion of the array in which the lower bound of a given value has to be looked for
+
+    end: int
+    The ending index of the portion of the array in which the lower bound of a given value has to be looked for
```

Comment:
The line width should not exceed 79 characters. Wrap up text in multiple lines not more than 79 chars